### PR TITLE
linkedList

### PR DESCRIPTION
**Pull request by pallav-jain-au3**

### CODE REVIEW

File: linkedList/doublyLL.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
+class DoublyLinkedList {
+    constructor(value) {
+        this.head = {
```
Comment (sharadJay) [inline linkedList/doublyLL.js:3]: Create a separate class for nodes

Comment (sharadJay) [inline linkedList/doublyLL.js:3]: there could be empty linked list as well, this is not possible with your implementation

File: linkedList/doublyLL.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            this.head = currentNode.next;
+            this.head.prev = null;
+            this.length--;
+            return
+        }
+        currentNode.prev.next = currentNode.next;
+        this.length--;
+    }
+
+    isEmpty() {
```
Comment (sharadJay) [inline linkedList/doublyLL.js:79]: make this as get property
```javascript
get isEmpty {
   return this.length === 0
}
```

File: linkedList/doublyLL.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        return this.length === 0;
+    }
+
+}
+let dll = new DoublyLinkedList(1);
```
Comment (sharadJay) [inline linkedList/doublyLL.js:84]: create tests in a separate file

File: linkedList/doublyLL.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            this.head = {
+                prev: null,
+                value,
+                next: null
+            };
+            this.tail = this.head;
+            this.length = 1;
+            return;
+        }
+        const node = {
```
Comment (sharadJay) [inline linkedList/doublyLL.js:22]: repetition of node creation

File: linkedList/doublyLL.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    contains(value) {
+        let currentNode = this.head;
+        while (currentNode !== null) {
+            if (currentNode.value === value) {
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
+        return false;
+
```
Comment (sharadJay) [inline linkedList/doublyLL.js:42]: remove space

**Review by sharadJay — CHANGES_REQUESTED** [review]

File: linkedList/linkedList.js [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+        this.tail = null;
+        this.length = 0;
+    }
+
+    //Inserting methods
+    addAtHead(value) {
+        if (this.isEmpty()) {
+            this.head = new ListNode(value);
+            this.tail = this.head;
+            this.length++;
```
Comment (sharadJay) [inline linkedList/linkedList.js]: this.length++;  is duplicate

File: linkedList/linkedList.js [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        this.length++;
+        return this.length;
+
+    }
+    addAtIndex(index, value) {
+        if (index > this.length) {
+            return;
+        } else if (index == this.length) {
+            return this.addAtTail(value);
+        } else if (index <= 0) {
```
Comment (sharadJay) [inline linkedList/linkedList.js:47]: if index is less than 0, node should not be added

File: linkedList/linkedList.js [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+            return this.addAtHead(value);
+        } else {
+            let node = new ListNode(value);
+            let currentNode = this.head;
+            let previousNode = null;
+            while (index) {
```
Comment (sharadJay) [inline linkedList/linkedList.js]: create a method for finding a node that could be used for next method as well

File: linkedList/middleNode.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+ *     this.val = val;
+ *     this.next = null;
+ * }
+ */
+/**
+ * @param {ListNode} head
+ * @return {ListNode}
+ */
+var middleNode = function(head) {
+    let lengthOfLinkedList = getLength(head);
```
Comment (sharadJay) [inline linkedList/middleNode.js:13]: remove comments

File: linkedList/removeNthFromLast.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
+
+var removeNthFromEnd = function(head, n) {
```
Comment (sharadJay) [inline linkedList/removeNthFromLast.js:2]: code on line 26 looks correct, why do we have 2 solutions?

File: linkedList/removeNthFromLast.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+        count++;
+    }
+    return count;
+}
+
+
+function removeNthNode(head ,n){
+    let previousNode = null;
+    let currentNode = head;
+    while(n){
```
Comment (sharadJay) [inline linkedList/removeNthFromLast.js:29]:
```
while(n > 0)
```

File: linkedList/linkedListProblem.js [hunk 12]
```diff
@@ -0,0 +1,51 @@
+
```
Comment (sharadJay) [inline linkedList/linkedListProblem.js:1]: make individual files for each implementation

File: Stack/isValid/isValid.js [hunk 13]
```diff
@@ -0,0 +1,24 @@
+let Stack = require("../Stack.js")
+var isValid = function(s) {
+      var map = {
+        '{':'}',
+        '(':')',
+        '[':']'
+    }
```
Comment (sharadJay) [inline Stack/isValid/isValid.js:7]: rename to brackets

File: Stack/isValid/isValid.js [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+let stack = new Stack();
+    for (let i = 0 ; i < s.length; i++){
+        if (s[i] == '[' || s[i] == '(' || s[i] == '{'){
```
Comment (sharadJay) [inline Stack/isValid/isValid.js]: can we utilize map here?
```
if map.keys().include(s[i])
```

File: Stack/isValid/isValid.js [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+            stack.push(s[i]);
+        }
+        else if (s[i] == ')' || s[i] == ']' || s[i] == '}'){
```
Comment (sharadJay) [inline Stack/isValid/isValid.js]: same can be done here

**Review by sharadJay — COMMENTED** [review]
added comments

File: linkedList/linkedList.js [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+            index--;
+        }
+        return currentNode.value;
+    }
+
+
+
+    //Checking value methods
+
+    isHead(node) {
```
Comment (sharadJay) [inline linkedList/linkedList.js:87]:
```
get isHead() {
   return this.head === node;
}
```
this.isHead

File: linkedList/linkedList.js [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+
+class LinkedList {
+    constructor() {
+        this.head = null;
+        this.tail = null;
+        this.length = 0;
+    }
+
+    //Inserting methods
+    addAtHead(value) {
```
Comment (sharadJay) [inline linkedList/linkedList.js:17]:
```
    addAtHead(value) {
      let newNode = new ListNode(value);
      if (this.isEmpty) {
        this.tail = this.head;
      } else {
        newNode.next = this.head;
      }
      this.head = newNode;
      this.length++;
      return this.length;
    }
```

File: linkedList/linkedList.js [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+        if (index > this.length) {
+            return;
+        } else if (index == this.length) {
+            return this.addAtTail(value);
+        } else if (index <= 0) {
+            return this.addAtHead(value);
+        } else {
+            let node = new ListNode(value);
+            let currentNode , previousNode ;
+           [previousNode, currentNode] = this.getPrevAndCurrNodesAtIndex(index);
```
Comment (sharadJay) [inline linkedList/linkedList.js]: func getNodeAtIndex(index)
prev= getNodeAtIndex(index- 1)
current = getNodeAtIndex(index)

File: linkedList/linkedList.js [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+        }
+        this.tail = currentNode;
+        currentNode.next = null;
+        this.length--;
+        return currentNode;
+    }
+
+    remove(value) {
+        if (this.isEmpty()) {
+            return undefined;
```
Comment (sharadJay) [inline linkedList/linkedList.js]: return

File: linkedList/linkedList.js [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+            return this.length++;
+        }
+    }
+    // get methods
+    get(index) {
+        if (index < 0 || index > this.length - 1) {
+            return -1;
+        }
+        let currentNode = this.head;
+        while (index) {
```
Comment (sharadJay) [inline linkedList/linkedList.js:64]: while index >= 0

File: linkedList/linkedList.js [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+            }
+            currentNode = currentNode.next;
+        }
+        return false;
+
+    }
+    //Removing node methods
+
+    removeHead() {
+        if (this.head.next == null) {
```
Comment (sharadJay) [inline linkedList/linkedList.js:111]: utilize removeAtIndex for removeHead, removeTail

File: linkedList/linkedList.js [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+            return this.removeTail();
+        } else {
+            previousNode.next = currentNode.next;
+            currentNode = null;
+            this.length--;
+            return;
+
+        }
+    }
+    getPrevAndCurrNodesAtIndex(index){
```
Comment (sharadJay) [inline linkedList/linkedList.js]: remove this method

**Review by sharadJay — CHANGES_REQUESTED** [review]
added comments

File: linkedList/linkedList.js (hunk 16 above)
Comment (pallav-jain-au3) [reply linkedList/linkedList.js:87]: this method is not for getting head value but to check that current node is head or not

Comment (sharadJay) [reply linkedList/linkedList.js:87]: right

**Review by sharadJay — APPROVED** [review]

**Pull request merged**